### ISSUE TITLE
Enhancement: prepare for the addition of `title` and `abstract` into `.entry.json` files

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/temp-test-ioccc#2671]
### Is there an existing issue for this?

- [X] I have searched for existing issues and did not find anything like this

### Describe the feature

Both `title` and `abstract`, according to [issue #979](https://github.com/ioccc-src/mkiocccentry/issues/979) in the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) are going to be added as mandatory fields to an entry's `.entry.json` file.  In addition, the value of `entry_JSON_format_version` will change.

This issue is to manage the impact of that change on this repo and the resulting web site.

### Relevant images, screenshots or other files

In [GH-issue-2546596062](https://github.com/ioccc-src/mkiocccentry/issues/979#issue-2546596062) we proposed that the `title` and `abstract` be added, in that order, after the `entry_id` and before the `author_set` as in:

```json
    "entry_id" : "1984_anonymous",
    "title" : "anonymous.1984",
    "abstract" : "prints hello world, where read is write",
    "author_set" : [
```

### Relevant links

See [issue #979](https://github.com/ioccc-src/mkiocccentry/issues/979) in the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry).

See also [summary.txt](https://www.ioccc.org/all/summary.txt) from the [Official IOCCC web site](https://www.ioccc.org/index.html).

### Anything else?

We propose the the so-callled "summary" be put into a JSON member called `abstract`.   While one could argue that "summary" might be a better name, keeping the word `abstract` is less disruptive to existing code where the JSON member `abstract` is already used for that purpose.    Keeping the name `abstract` takes less time.  Moreover the actual JSON member name is less important than how the JSON member value is used for the web site.

# TODOs

* [x] Update the `entry_JSON_format_version` value as needed in all entry `.entry.json` files

* [x] Add `title` to in all entry `.entry.json` files

We recommend as a default, to as the default `title` JSON member value: **dirname.YYYY**.

 **IMPORTANT NOTE:** The above is just a suggest default.  The `title` JSON member value does **NOT** need to be in that form.  For example, it could be:

```json
    "entry_id" : "1984_anonymous",
    "title" : "hello.read.as.write",
    "abstract" : "prints hello world, where read is write",
    "author_set" : [
```

Understand that the `title` might not be used in the web site right away.  Future comments on this issue will likely discuss.  We are simply putting the `title` form the `.info.json` file into an entry's `.entry.json` file.  We also simply propose a simple way to initialize `title` JSON member value.

**SUGGESTION:** Do **NOT** worry about the format for default for now.  Set the default for now.  The `title` might not be used to form web pages: it is there in case.  And if there some use for the `title` comes along, the `title`  JSON member values can be changed as needed based in the use case.

* [x] Add `abstract` to in all entry `.entry.json` files

We recommend as a default, to use the contents of [summary.txt](https://www.ioccc.org/all/summary.txt) from the [Official IOCCC web site](https://www.ioccc.org/index.html) for the JSON member value.  Feel free to improve on the text as needed.

Keep in mind that the limits imposed on the ["other repo"](https://github.com/ioccc-src/mkiocccentry) on the must apply to the new `abstract`.   Because [issue 940 form the "other repo"](https://github.com/ioccc-src/mkiocccentry/issues/940) won't be ready for some time, adherence to the limits imposed will have to be _manually enforced_ for now.

**SUGGESTION:** Do **NOT** worry about default value of `abstract` for now.  Start with the suggested [summary.txt](https://www.ioccc.org/all/summary.txt) from the [Official IOCCC web site](https://www.ioccc.org/index.html) for now.  Wait to see how the `abstract` is used in HTML pages before making major changes.

* [x] Update FAQ documentation

Where need, update `faq.md` to include the result of [issue #979](https://github.com/ioccc-src/mkiocccentry/issues/979) in the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry).

**IMPORTANT:** When documenting the `title`, indicate that this field comes from the same named JSON member found in submission `.info.json` files.  Indicate that for entries prior to 2024, the field was initialized as **dirname.YYYY** just to get stared, however the `title` does **NOT** have to be of that form **AND** is the field is used in the future, the `title` for older entries could be subject to change.

According to `test_title()` in [soup/entry_util.c from the "other repo"](https://github.com/ioccc-src/mkiocccentry/blob/master/soup/entry_util.c), the title must be between 1 and `MAX_TITLE_LEN` characters that pass the `posix_plus_safe(str, true, false, true)` test.  Obviously the would need to explain in plain English, the effect of the test.  Use the current value of `MAX_TITLE_LEN` in the FAQ instead of the symbol name in the FAQ.

**IMPORTANT:** When documenting the `abstract`, explain that that for entries prior to 2024, the historic contents of the `summary.txt` file was used.  We suggest a wayback machine link be used as after the **Great Fork Merge** that file will go away.

According to `test_abstract()` in [soup/entry_util.c from the "other repo"](https://github.com/ioccc-src/mkiocccentry/blob/master/soup/entry_util.c), the abstract must be between 1 and `MAX_ABSTRACT_LEN` characters.  Use the current value of `MAX_ABSTRACT_LEN ` in the FAQ instead of the symbol name in the FAQ.

Feel free to describe the JSON member name, `abstract` as a historical artifact and that one may consider the JSON member value to be sort of a very short summary.

We do plan to use the `abstract` when building HTML pages   How exactly is TBD.

* [x] update bin tools

A number of the current `bin/` tools will need to be updated as a result of this issue.

Initially the `bin/` tools will be "forgiving" of a lack of a `title` and `abstract`.

Once it is determined how the `abstract` will be used to build HTML pages, the "**work-around for missing `title` and `abstract`"** can be adjusted.

We plan to make the  `bin/` tools  be adaptive to the lack of a `title` and `abstract` soon.

* [x] determine how the `abstract` will be used to build HTML pages

We suggest a disccion below in this issue would be helpful.  Keep in mind, however the existing limits `test_abstract()` in [soup/entry_util.c from the "other repo"](https://github.com/ioccc-src/mkiocccentry/blob/master/soup/entry_util.c) need to remain in place.

We will "UPDATE 0" as needed.

Comments welcome.

## UPDATE 0

**USEFUL HINT:** To make use of a CSV file easier, do not use a comma ("_,_") in the `abstract`.

**Comment by lcn2** [comment]
With commit c1f2a5f5e2f187dff45de69b20f4f0494d0790c3 we added a temporary `summary.csv` file that is based on the [summary.txt](https://github.com/ioccc-src/winner/blob/master/all/summary.txt).

We converted the contents into a CSV format.  Moreover the abstract was limited to no more than 63 characters.  All comma's (",") within the abstract were removed to make converting to CSV easier.

NOTE: This is a tempory file that will **GO AWAY** once the all of the `.entry.json` files have been updated with a title and abstract.

**Comment by xexyl** [comment]
Please assign this to me and I hope to look into it soon .. after the changes in the other repo that are needed are done.

**Comment by lcn2** [comment]
> Please assign this to me and I hope to look into it soon .. after the changes in the other repo that are needed are done.

We will, however, take care of the bin tools. We will also pre-load the `.entry.json` files with the data from the temporary `summary.csv` file. 

Please wait until that process is ready.

**Comment by xexyl** [comment]
Thanks! Let me know when you'd like me to work on this then, please.

**Comment by lcn2** [comment]
We are nearly finished ... testing ... perhaps by the end of Friday the bin tools and the `.entry.json` files will be ready.

**Comment by lcn2** [comment]
With commit 0fa01fd7441baf8bf695c69bb1829e4450f989f4 a "_title_" and "_abstract_" has been added to all `.entry.json` files.

**NOTE:** the "_title_" is not current used (and may not be used), but is there if needed.  The "_title_" defaults to "_YYYY.dir_".

The "_abstract_" is used to add a `<h3>abstract text</h3>` to entry `index.html` files.

The `make entry2csv` rule will create a temporary `summary.csv` file there the title and abstract for each entry are listed. 

You may modify the `summary.csv` file  as needed.  Once modified, the `make csv2entry` rule will update `.entry.json` files as needed.

Keep in mind that the "_abstract_" must be <= 64 characters in length and may not contain a ; (semicolon), nor & (ampersand), nor , (comma).

TODO: Update documentation.

**Comment by lcn2** [comment]
We believed we have completed the actions of this issue.

Comments welcome.

**Comment by xexyl** [comment]
> We have completed the all actions of this issue and that it is ready to be closed.
> 
> Sure, some of the "_abstract_" strings for some entries could be improved, but that is more of an issue #2006 so we will not address it here.

Hmm ... I hope I do not have to look at all the entries again!

But for the abstract I think the answer is simple: it should be the strings in `summary.txt`. That was my original idea AND it's what was done in some entries too. If this was done that way then I do not see the problem. If not how can we fix it? Maybe a simple script (or a more complicated script?) would update the strings?

If we did this then a script with a JSON extractor (eventually ours but something else in the meantime) could make it a html file which would be a good thing I think (the file being summary.txt to summary.html).

**Comment by lcn2** [comment]
> > We have completed the all actions of this issue and that it is ready to be closed.
> > Sure, some of the "_abstract_" strings for some entries could be improved, but that is more of an issue #2006 so we will not address it here.
> 
> Hmm ... I hope I do not have to look at all the entries again!
> 
> But for the abstract I think the answer is simple: it should be the strings in `summary.txt`. That was my original idea AND it's what was done in some entries too. If this was done that way then I do not see the problem. If not how can we fix it? Maybe a simple script (or a more complicated script?) would update the strings?
> 
> If we did this then a script with a JSON extractor (eventually ours but something else in the meantime) could make it a html file which would be a good thing I think (the file being summary.txt to summary.html).

The were the strings in the `summary.txt` file, except for a few with some problematic characters for CSV and for HTML encoding and shell quoting.  So we had to make a few edits.

**Comment by lcn2** [comment]
> Hmm ... I hope I do not have to look at all the entries again!

if you do want to review the abstract strings, just build the `summary.csv` file:

```sh
make entry2csv
```

Then open the first entry:

```sh
open 1984/anonymous/index.html
```

Look at the abstract printed just above the navbar line and click on the next entry ... which now can go all the way through the entry set.

Modify anything you want to improve in  the `summary.csv` file, and then run the update command:

```sh
make csv2entry
```

It is really easy to step through.

**Comment by xexyl** [comment]
Re the above two comments: it seems good. As long as it can be used to form the summary.html file I don't see any problem.

Thanks for the details! We just have to write a script now that can form the summary.html file.

Will reply to other things later ... working on is available stuff in jparse: that way I can do it one commit here too. But have to go afk a bit.
